File: src/main/kotlin/magicbook/gtlitecore/common/item/behavior/HaloRenderItemBehavior.java
package magicbook.gtlitecore.common.item.behavior;

import codechicken.lib.model.ModelRegistryHelper;
import codechicken.lib.util.TransformUtils;
import magicbook.gtlitecore.client.renderer.IHaloRenderBehavior;
import magicbook.gtlitecore.client.renderer.handler.HaloItemRenderer;
import net.minecraft.client.renderer.block.model.ModelResourceLocation;
import net.minecraft.client.renderer.texture.TextureAtlasSprite;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import java.util.function.Supplier;

public class HaloRenderItemBehavior implements IHaloRenderBehavior
{

    private final int haloSize;
    private final int haloColor;
    private final Supplier<?> supplier;
    private final boolean drawPulse;

    public HaloRenderItemBehavior(int haloSize, int haloColor,
                                  Supplier<?> supplier, boolean drawPulse)
    {
        this.haloSize = haloSize;
        this.haloColor = haloColor;
        this.supplier = supplier;
        this.drawPulse = drawPulse;
    }

    @Override
    public boolean shouldDrawHalo()
    {
        return true;
    }

    @SideOnly(Side.CLIENT)
    @Override
    public TextureAtlasSprite getHaloTexture()
    {
        return (TextureAtlasSprite) this.supplier.get();
    }

    @Override
    public int getHaloColor()
    {
        return this.haloColor;
    }

    @Override
    public int getHaloSize()
    {
        return this.haloSize;
    }

    @Override
    public boolean shouldDrawPulse()
    {
        return this.drawPulse;
    }

    @SideOnly(Side.CLIENT)
    @Override
    public void onRendererRegistry(ResourceLocation location)
    {
        ModelRegistryHelper.register(new ModelResourceLocation(location, "inventory"),
                new HaloItemRenderer(TransformUtils.DEFAULT_ITEM, modelRegistry
                        -> modelRegistry.getObject(new ModelResourceLocation(location, "inventory"))));
    }

}

File: src/main/kotlin/gregtechlite/gtlitecore/common/item/behavior/HaloRenderItemBehavior.kt
package gregtechlite.gtlitecore.common.item.behavior

import codechicken.lib.model.ModelRegistryHelper
import codechicken.lib.util.TransformUtils
import gregtechlite.gtlitecore.client.renderer.HaloRenderBehavior
import gregtechlite.gtlitecore.client.renderer.handler.item.HaloItemRenderer
import net.minecraft.client.renderer.block.model.ModelResourceLocation
import net.minecraft.client.renderer.texture.TextureAtlasSprite
import net.minecraft.util.ResourceLocation
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly

class HaloRenderItemBehavior(private val haloSize: Int,
                             private val haloColor: Int,
                             private val haloTexture: () -> () -> TextureAtlasSprite,
                             private val drawPulse: Boolean) : HaloRenderBehavior
{
    override fun shouldDrawHalo(): Boolean = true

    @SideOnly(Side.CLIENT)
    override fun getHaloTexture(): TextureAtlasSprite? = haloTexture.invoke().invoke()

    override fun getHaloColor(): Int = haloColor

    override fun getHaloSize(): Int = haloSize

    override fun shouldDrawPulse(): Boolean = drawPulse

    @SideOnly(Side.CLIENT)
    override fun onRendererRegistry(location: ResourceLocation)
    {
        ModelRegistryHelper.register(ModelResourceLocation(location, "inventory"),
            HaloItemRenderer(TransformUtils.DEFAULT_ITEM) {
                it.getObject(ModelResourceLocation(location, "inventory"))
            })
    }
}